app/scripts/lib/transaction-metrics.ts

```diff
@@ -0,0 +1,879 @@
+import { isHexString } from 'ethereumjs-util';
+import EthQuery from 'eth-query';
+import { BigNumber } from 'bignumber.js';
+import type { Provider } from '@metamask/network-controller';
+
+import { ORIGIN_METAMASK } from '../../../shared/constants/app';
+import {
+  determineTransactionAssetType,
+  isEIP1559Transaction,
+} from '../../../shared/modules/transaction.utils';
+import { hexWEIToDecGWEI } from '../../../shared/modules/conversion.utils';
+import {
+  TransactionType,
+  TokenStandard,
+  TransactionApprovalAmountType,
+  TransactionMetaMetricsEvent,
+  TransactionMeta,
+} from '../../../shared/constants/transaction';
+import { MetaMetricsEventCategory } from '../../../shared/constants/metametrics';
+import { GasRecommendations } from '../../../shared/constants/gas';
+import { TRANSACTION_ENVELOPE_TYPE_NAMES } from '../../../shared/lib/transactions-controller-utils';
+
+///: BEGIN:ONLY_INCLUDE_IN(blockaid)
+import {
+  BlockaidReason,
+  BlockaidResultType,
+} from '../../../shared/constants/security-provider';
+///: END:ONLY_INCLUDE_IN
+
+export const METRICS_STATUS_FAILED = 'failed on-chain';
+
+type ControllerActions = {
+  createEventFragment: (arg0: any) => any;
+  finalizeEventFragment: (arg0: string, arg1?: any) => any;
+  getEventFragmentById: () => any;
+  updateEventFragment: (arg0: string, arg1: any) => any;
+  getAccountType: (arg0: string) => Promise<any>;
+  getDeviceModel: (arg0: string) => Promise<any>;
+  getEIP1559GasFeeEstimates: () => any;
+  getSelectedAddress: () => any;
+  getTokenStandardAndDetails: () => any;
+  getTransaction: (arg0: string) => TransactionMeta;
+};
+
+type TrackerDependencyMap = {
+  controllerActions: ControllerActions;
+  provider: Provider;
+};
+
+// transaction-added event doesn't have to update/finalize fragment
+export const onTransactionAdded =
```

Comment:
Is it best practice to add JSDoc to these functions since they're exported?

Reply:
Good call, sorted in https://github.com/MetaMask/metamask-extension/pull/21330/commits/a5a32f8ed27f3f3fd223c3e07fa5c81b011bd409